### DISCUSSION TITLE
Subscribe to Shared state updates outside of Reducer

### DISCUSSION THREAD

**Discussion by atacan** [pointfreeco/swift-composable-architecture#3101] [Q&A]
Is there a way to subscribe to the changes that happen to a shared state outside of a reducer, particularly the state that is persisted on file storage? 

It would be great if I could subscribe to changes within a dependency to store them in its `LockIsolated` value.

```swift
extension ApiClient: DependencyKey {
    public static var liveValue: Self {
        // I want to keep this up to date
        let mydata = ActorIsolated(ApiClientData())

        return Self(
            startTask: { fileUrl in
                // use mydata.value
            }
    }
}
```

When a user makes changes on the settings window, for example, I couldn't expose an endpoint on the client to update its internal parameters, because the settings module did not need to import the dependency's module, and the dependency needs a long build time.

I guess the solution right now is to import it and do the following in the settings reducer?

```swift
case .onAppear: // or simply case .binding(\...)
  return .publisher {
    state.$count.publisher
      .map(Action.countUpdated)
  }
case .countUpdated(let count):
  // ...
  apiClient.updateCount(count)
  // ...
```

**Comment by mbrandonw** [comment]
> Is there a way to subscribe to the changes that happen to a shared state outside of a reducer, particularly the state that is persisted on file storage?

You can use `@Shared` and `$shared.publisher` in your dependency. You may not even need the isolated value in your live dependency.

**Comment by atacan** [comment]
Thank you for your help. I couldn't make it work. Here I made a toy project. https://github.com/atacan/DiscusssionSharedInDependency/blob/main/DiscusssionSharedInDependency/ContentView.swift

I used `@Shared` in the `liveValue` property.  
I am not sure how or where to subscribe to that publisher.  
It also doesn't fetch the file content even when the app restarts. The file is changed and restored by the view and the reducer correctly.

**Reply by mbrandonw** [reply]
Nothing is holding onto `cancellables` and so it is deallocated immediately and the subscription is torn down. You could capture in the `get` closure to keep it alive as long as that closure is around.
